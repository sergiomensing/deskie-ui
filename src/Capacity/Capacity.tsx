import { HTMLAttributes, PropsWithChildren } from 'react';
import styled, { css } from 'styled-components';
import { Text } from '../index';

type CapacitySize = 'default' | 'large';
type CapacityProps = HTMLAttributes<unknown> & {
  size?: CapacitySize;
  capacity: number;
};

const StyledCapacity = styled.div<{ labelInside: boolean }>`
  display: flex;
  align-items: center;

  .capacity-label {
    font-size: 0.75rem;
    margin-right: 0.25rem;
  }

  ${(props) =>
    props.labelInside &&
    css`
      position: relative;
      justify-content: center;

      .capacity-label {
        position: absolute;
        margin: 0;
        text-align: center;
      }
    `}
`;

function getStokeColor(capacity: number) {
  if (capacity === 100) {
    return '#ff754c';
  }
  if (capacity >= 80) {
    return '#ffce73';
  }

  return '#7fba7a';
}

function getSize(size: CapacitySize) {
  if (size === 'large') {
    return { stroke: 4, radius: 24 };
  }
  return { stroke: 4, radius: 14 };
}

function Capacity({ children, capacity, size = 'default' }: PropsWithChildren<CapacityProps>) {
  const _capacity = capacity > 100 ? 100 : capacity;
  const { radius, stroke } = getSize(size);
  const normalizedRadius = radius - stroke;
  const circumference = normalizedRadius * 2 * Math.PI;
  const strokeDashoffset = circumference - (_capacity / 100) * circumference;

  return (
    <StyledCapacity labelInside={size === 'large'} className="capacity">
      {children && (
        <Text className="capacity-label" small appearance={size === 'large' ? 'default' : 'faded'}>
          {children}
        </Text>
      )}
      <svg className="capacity-ring" height={radius * 2} width={radius * 2} style={{ transform: 'rotate(-90deg)' }}>
        <circle stroke="#eeeeee" fill="transparent" strokeWidth={stroke} r={normalizedRadius} cx={radius} cy={radius} />
        <circle
          stroke={getStokeColor(_capacity)}
          strokeLinecap="round"
          fill="transparent"
          strokeWidth={stroke}
          strokeDasharray={circumference + ' ' + circumference}
          style={{ strokeDashoffset }}
          r={normalizedRadius}
          cx={radius}
          cy={radius}
        />
      </svg>
    </StyledCapacity>
  );
}

export default Capacity;
