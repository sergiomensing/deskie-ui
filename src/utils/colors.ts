export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
