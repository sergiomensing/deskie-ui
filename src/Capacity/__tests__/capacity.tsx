import React from 'react';
import Capacity from '../Capacity';
import { render } from '../../../tests/utils';

describe('Capacity', () => {
  it('should support all sizes', () => {
    const { container } = render(
      <>
        <Capacity capacity={50} size="default" />
        <Capacity capacity={50} size="large" />
      </>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should fill svg ring with capacity value', () => {
    const { container } = render(
      <>
        <Capacity capacity={0} />
        <Capacity capacity={20} />
        <Capacity capacity={80} />
        <Capacity capacity={100} />
      </>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render children for all sizes', () => {
    const { getAllByText } = render(
      <>
        <Capacity capacity={20}>test-value</Capacity>
        <Capacity size="large" capacity={20}>
          test-value
        </Capacity>
      </>,
    );
    expect(getAllByText('test-value').length).toBe(2);
  });
});
