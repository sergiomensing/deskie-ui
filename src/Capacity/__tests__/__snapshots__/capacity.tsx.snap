// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Capacity should fill svg ring with capacity value 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .capacity-label {
  font-size: 0.75rem;
  margin-right: 0.25rem;
}

@media screen and (min-width:1024px) {

}

<div>
  <div
    class="c0 capacity"
  >
    <svg
      class="capacity-ring"
      height="28"
      style="transform: rotate(-90deg);"
      width="28"
    >
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#eeeeee"
        stroke-width="4"
      />
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#7fba7a"
        stroke-dasharray="62.83185307179586 62.83185307179586"
        stroke-linecap="round"
        stroke-width="4"
        style="stroke-dashoffset: 62.83185307179586;"
      />
    </svg>
  </div>
  <div
    class="c0 capacity"
  >
    <svg
      class="capacity-ring"
      height="28"
      style="transform: rotate(-90deg);"
      width="28"
    >
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#eeeeee"
        stroke-width="4"
      />
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#7fba7a"
        stroke-dasharray="62.83185307179586 62.83185307179586"
        stroke-linecap="round"
        stroke-width="4"
        style="stroke-dashoffset: 50.26548245743669;"
      />
    </svg>
  </div>
  <div
    class="c0 capacity"
  >
    <svg
      class="capacity-ring"
      height="28"
      style="transform: rotate(-90deg);"
      width="28"
    >
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#eeeeee"
        stroke-width="4"
      />
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#ffce73"
        stroke-dasharray="62.83185307179586 62.83185307179586"
        stroke-linecap="round"
        stroke-width="4"
        style="stroke-dashoffset: 12.566370614359172;"
      />
    </svg>
  </div>
  <div
    class="c0 capacity"
  >
    <svg
      class="capacity-ring"
      height="28"
      style="transform: rotate(-90deg);"
      width="28"
    >
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#eeeeee"
        stroke-width="4"
      />
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#ff754c"
        stroke-dasharray="62.83185307179586 62.83185307179586"
        stroke-linecap="round"
        stroke-width="4"
        style="stroke-dashoffset: 0;"
      />
    </svg>
  </div>
</div>
`;

exports[`Capacity should support all sizes 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .capacity-label {
  font-size: 0.75rem;
  margin-right: 0.25rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 .capacity-label {
  font-size: 0.75rem;
  margin-right: 0.25rem;
}

.c1 .capacity-label {
  position: absolute;
  margin: 0;
  text-align: center;
}

@media screen and (min-width:1024px) {

}

<div>
  <div
    class="c0 capacity"
  >
    <svg
      class="capacity-ring"
      height="28"
      style="transform: rotate(-90deg);"
      width="28"
    >
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#eeeeee"
        stroke-width="4"
      />
      <circle
        cx="14"
        cy="14"
        fill="transparent"
        r="10"
        stroke="#7fba7a"
        stroke-dasharray="62.83185307179586 62.83185307179586"
        stroke-linecap="round"
        stroke-width="4"
        style="stroke-dashoffset: 31.41592653589793;"
      />
    </svg>
  </div>
  <div
    class="c1 capacity"
  >
    <svg
      class="capacity-ring"
      height="48"
      style="transform: rotate(-90deg);"
      width="48"
    >
      <circle
        cx="24"
        cy="24"
        fill="transparent"
        r="20"
        stroke="#eeeeee"
        stroke-width="4"
      />
      <circle
        cx="24"
        cy="24"
        fill="transparent"
        r="20"
        stroke="#7fba7a"
        stroke-dasharray="125.66370614359172 125.66370614359172"
        stroke-linecap="round"
        stroke-width="4"
        style="stroke-dashoffset: 62.83185307179586;"
      />
    </svg>
  </div>
</div>
`;
